### ISSUE TITLE
RUST-36 Support of Decimal128 type

### ISSUE THREAD

**Issue by zonyitoo** [mongodb/bson-rust#53]
According to [BSON 1.1 Spec](http://bsonspec.org/spec.html), it introduced a new type `decimal128`.

But currently Rust doesn't support `d128` type natively.

Related links

1. `d128` crate: https://github.com/alkis/decimal
2. Official support RFC: https://github.com/rust-num/num/issues/8

**Comment by zonyitoo** [comment]
Thoughts? @kyeah

**Comment by kyeah** [comment]
nice, just one comment; looks great otherwise.

**Comment by saghm** [comment]
I believe crates.io doesn't let you publish a crate that has both git and crates.io dependencies. Is there any chance that the author of the decimal crate might accept a PR?

**Comment by zonyitoo** [comment]
I have already submitted a PR to https://github.com/alkis/decimal/pull/23 . I will remove that git dep after this PR is accepted.

**Comment by zonyitoo** [comment]
Don't know if it is the best way to support decimal 128 in Rust, yet.

Rust is now supported `i128` and `u128`. Maybe in the future it will support `d128`?

**Comment by lrlna** [comment]
@zonyitoo by `the best way` do you mean you're not sure if using the `decimal` crate is a good idea, or is this more about whether the support is up to the spec? 

it's possible core will support `d128`, but I feel like that might be a long way off :(. In the meantime would be cool to have a workaround \o/

**Comment by zonyitoo** [comment]
Yes, I agree. I would take sometime to make this commit works with the master.

**Comment by lrlna** [comment]
@zonyitoo if you feel like it's more work than you have capacity for, I can also help out! I just didn't want to step on your toes, since there was already a commit going.

**Comment by zonyitoo** [comment]
@lrlna That would be great! I don't have spear time in the recent future :(

**Comment by saghm** [comment]
We can use https://jira.mongodb.org/browse/RUST-36 to track the work done for this issue

**Comment by nevi-me** [comment]
I filed https://github.com/mongodb-rust/mongodb-schema-parser/pull/35 before I saw this conversation. I am trying to parse a collection's schema using the schema parser. However because `mongodb` doesn't build with the `feature = ["decimal128"]` from the `bson` crate, I'm encountering a non-exhaustive pattern as per below.

```rust
    Checking mongodb v0.9.1
error[E0004]: non-exhaustive patterns: `&Decimal128(_)` not covered
   --> C:\Users\me\.cargo\registry\src\github.com-1ecc6299db9ec823\mongodb-0.9.1\src\bson_util.rs:128:11
    |
128 |     match val {
    |           ^^^ pattern `&Decimal128(_)` not covered
    |
    = help: ensure that all possible cases are being handled, possibly by adding wildcards or more match arms
```

This is with the below Cargo.toml

```toml
[dependencies]
bson = { version = "0.14", features = ["decimal128"] }
mongodb = "0.9.1"
# this contains the patch to use bson 0.14 on the schema parser
mongodb-schema-parser = { git = "https://github.com/nevi-me/mongodb-schema-parser", branch = "update-bson" }
```

I don't need decimal128 support, so I'll fork the `mongodb` crate and add the extra pattern for now.

**Comment by bugproof** [comment]
I tried to read data from mongodb collection that uses decimal128 but it panics

> thread 'main' panicked at 'Attempted conversion of invalid data type: Decimal128(...)', C:\Users\Konrad\.cargo\registry\src\github.com-1ecc6299db9ec823\bson-2.0.0-beta.1\src\bson.rs:587:18
stack backtrace:...

Is there any workaround or it's simply not possible with rust's mongodb library?

**Comment by ale-rinaldi** [comment]
Hello,

I'm having a fairly serious issue in using the decimal128 feature, with bson 2.0.0-beta.2.

I created a really simple, compilable project that describes the issue: https://github.com/ale-rinaldi/rust-decimal-test

The important part is here:
```rust
/* uses */

#[derive(Serialize)]
struct TestData {
    pub(crate) my_date: bson::DateTime,
    pub(crate) my_decimal: bson::Decimal128,
    pub(crate) my_string: String,
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    /* Code for connection */

    let data = TestData {
        my_date: bson::DateTime::from_millis(1278720000000),
        my_decimal: bson::Decimal128::from_u32(42),
        my_string: "foobar".to_string()
    };
    let doc = bson::to_document(&data).unwrap();
    collection.insert_one(doc, None).await.unwrap();
    Ok(())
}
```

So I'm creating a simple test struct with a Date, a Decimal128 and a String, and saving them to Mongo (I'm on MongoDB 4.1.13).
What happens here is that, if I look at the created document using MongoDB Compass, the date and the string fields have the value that I except (`2010-07-10T00:00:00.000+00:00` and `foobar`), while the decimal has no value 42 but `6.6E-1819`.

The strange thing is that, if I read the document back from Rust, the value of the returned Decimal128 is now correct (42).

Am I totally missing the point here, or is it an actual problem with the `bson` serializer?

Thanks

**Comment by patrickfreed** [comment]
@bugproof This was a bug actually that has since been fixed in `2.0.0-beta.3` and `1.2.3`. Note that the `decimal128` feature flag has been removed in `2.0.0-beta.3` (see #282), so you'll need to remove that workaround. Deserialization from documents that contain decimal128 values should work just fine now though.

**Comment by dotjon0** [comment]
Just checking in to see if there has been any update on this?

**Comment by dotjon0** [comment]
Could the MongoDB team provide feedback of where support for Decimal128 is at? This issue was originally raised in 2017 and with no recent updates. 

A could of key points:

1. MongoDB recommends Decimal128 as the best option for 'monetary' high precision values https://www.mongodb.com/docs/manual/tutorial/model-monetary-data/ 

2. Rust is becoming a very mainstream language. Lack of Decimal128 support could block development teams from using MongoDB Atlas if they have adopted rust.

**Comment by kmahar** [comment]
Hi @dotjon0, our current plan is to start implementation of this ~1 month from now.

**Comment by dotjon0** [comment]
@kmahar thank you so much for the update, its very appreciated. Great to hear Decimal128 implementation is commencing in only a few weeks! Absolute fantastic news! Will look forward and thank you for making the time to update us.

**Comment by lrlna** [comment]
@kmahar hello 👋 ! I think there might be some interest in rust's standard library to have rust support decimal 128 implementation natively. If you end up writing this implementation from scratch, would be super cool if you could contribute it back to rust itself :) There is already a [`u128`](https://doc.rust-lang.org/std/primitive.u128.html)/[`i128`](https://doc.rust-lang.org/std/primitive.i128.html), so `std` is close!

**Comment by kmahar** [comment]
hey @lrlna! 🙂 I think to start this type would likely only support converting back and forth between string <-> decimal128 byte representations, and in the future we may try to support conversion to/from other Rust numeric types. If there was a `std` decimal128 type we’d want to support interop with that, and we could potentially switch to using it in lieu of our type if it supported BID coefficient encoding.
We’d be very happy to contribute our string <-> bytes logic upstream if that helps along an implementation, but at this time I think it’s unlikely we’ll end up writing a full-featured type ourselves that directly supports math like I assume a `std` type would. However, if that changes we’ll definitely keep this in mind!

**Comment by dotjon0** [comment]
> hey @lrlna! 🙂 I think to start this type would likely only support converting back and forth between string <-> decimal128 byte representations, 

amazing, exactly what we need! Thanks so much @kmahar

**Comment by abr-egn** [comment]
Sorry for the long delay here, just wanted to let people know that implementation of conversion to and from human-readable strings for the `Decimal128` type has started, and we expect this to go out in the next driver release.  If there's other functionality that would be particularly useful, please let us know!  It's unlikely to be part of the upcoming feature addition but will definitely inform future priorities.

**Comment by abr-egn** [comment]
We just released version 2.6.0 of the bson crate, which includes support for conversion between Decimal128 values and human-readable strings 🙂

**Comment by dotjon0** [comment]
> We just released version 2.6.0 of the bson crate, which includes support for conversion between Decimal128 values and human-readable strings 🙂

amazing @abr-egn thank you so much!
